Group field-caps node requests by index mapping hash

dnhatn opened the pull request:
This optimization is for field-caps requests targeting many indices with an index pattern. Instead of reaching out to many data nodes to retrieve field-caps, we can group indices by their mapping hashes then send a single node request with representative indices instead. This optimization is significant in large clusters.

dnhatn committed:
Group field-caps node requests by index mapping hash

dnhatn committed:
Fix tests

dnhatn committed:
Fix more tests

dnhatn committed:
wording

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into field_caps_groups

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

elasticsearchmachine commented:
Hi @dnhatn, I've created a changelog YAML for you.

dnhatn committed:
Update docs/changelog/84598.yaml

javanna commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -191,67 +194,111 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
+        }
+        for (Map.Entry<String, List<String>> e : withMappingHashes.entrySet()) {
+            final GroupShardsIterator<ShardIterator> shardIts = clusterService.operationRouting()
+                .searchShards(clusterState, e.getValue().toArray(String[]::new), null, null, null, null);
```
question: if the mapping is the same for some indices, can we just ask field_caps for a single index and apply what we get to all the other indices with the same hash?

javanna commented:
Do I understand correctly that before we would ask for field_caps for all indices, and eventually get back one set of fields per distinct mapping hash; with this change, we send one single request per mapping hash, and then apply the same response to all indices that have the same hash? This has the advantage of minimizing the amount of roundtrips when many indices have the same mappings: many indices all with the same mappings would likely send one request per data node, and get back the full set of fields once from each node. This is much better than what we did before as we would otherwise repeat the set of fields for every index, resulting in much bigger transport response. Though with this change, we would ask only one node, and get back the full set of fields once from it, so the amount of requests is no longer a function of the number of data nodes (again provided that mappings are the same for all indices involved).

This change is a great improvement, but it would be even better to be able to measure the improvement through benchmarks, relates to #84504  .

@original-brownbear would you mind having a look too given your involvement with the many shards scalability project?

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into field_caps_groups

jtibshirani commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -191,67 +194,111 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
 
-    private void afterRequestsCompleted(int numRequests) {
-        if (pendingRequests.addAndGet(-numRequests) == 0) {
+    private void afterRequestsCompleted() {
```
Tiny comment, I think this should be singular `afterRequestCompleted`.

jtibshirani commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -56,15 +59,15 @@ final class RequestDispatcher {
     private final long nowInMillis;
 
-    private final boolean hasFilter;
+    private final boolean withIndexFilter;
```
I think it'd be helpful to describe the execution strategy in the  javadoc for the `RequestDispatcher` class, because it would really helps others in understanding this logic. We could summarize how we deduplicate based on mapping hash, and how we need to disable the optimizations when there's an index filter.

dnhatn replied:
+1, I added a javadocs in https://github.com/elastic/elasticsearch/pull/84598/commits/886893e0654bf7f79376b04f8a4d70d6cd433afa

jtibshirani reviewed (commented):
This looks like a great optimization! I left a few small comments but the logic + tests look good to me.

I wonder if Kibana is using `index_filter` or not, this will really affect when the optimization helps! Unfortunately I think it would require a big change to how we execute field caps in order to get it to work with `index_filter`...

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into field_caps_groups

dnhatn committed:
HashMap

dnhatn committed:
Use shorted version

dnhatn committed:
Do not send requests for empty indices

dnhatn committed:
javadocs

dnhatn committed:
add mapping

original-brownbear commented:
Sorry for the delay here Nhat! I'm looking at this today and will try to at least manually benchmark this a little.

original-brownbear commented:
@dnhatn I benchmarked this against the many shards benchmark setup now. Interestingly it does not provide any (measurable) throughput improvement. My best guess as to why that is, is that we are bottle-necked on the REST side of the network layer in some form (this isn't something that can be fixed here so I wouldn't worry about this). 
It does however look like this change measurably reduces the GC pressure (about half as much CPU goes into GC after these changes) and while I was not able to measure the effects of this on the amount of bytes+messages that the transport layer has to handle (not because they're not there, just a matter of the complexity of setting up the experiment), reducing them is always a win in terms of stability.

The before and after of `innerMerge` (which is effectively all the CPU load on the coordinating node during field_caps) also looks nicer now and is able to compile the capturing lambda `key -> new FieldCapabilities.Builder(field, key)` better which saves us some CPU on the coordinator. 

before:

![image](https://user-images.githubusercontent.com/6490959/165732289-b1737989-8301-44c4-a9e3-2e1cff732e8b.png)


after:

![image](https://user-images.githubusercontent.com/6490959/165732207-16a258ed-c597-4865-88c3-e18cf41db26f.png)


-> LGTM from my end though I agree with Julie's points on documentation :)

original-brownbear reviewed (commented):
see above

dnhatn commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -191,67 +194,111 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
+        for (String index : indices) {
+            final IndexMetadata indexMetadata = clusterState.metadata().index(index);
+            if (withIndexFilter == false
+                && indexMetadata != null
```
I added a comment and skip requests for empty indices: https://github.com/elastic/elasticsearch/pull/84598/commits/68d6ddd3536177ed638dad93e69f7624f3ee8439

dnhatn commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -191,67 +194,111 @@ private void sendRequestToNode(String nodeId, List<ShardId> shardIds) {
+            } else {
+                final GroupShardsIterator<ShardIterator> shardIts = clusterService.operationRouting()
+                    .searchShards(clusterState, new String[] { index }, null, null, null, null);
+                groups.add(new Group(List.of(index), null, shardIts));
```
Good catch. Not sure why I used a long version. See https://github.com/elastic/elasticsearch/pull/84598/commits/96210f7c94d71e5ee611b88443bb75289935e952.

dnhatn committed:
HashSet

dnhatn commented on server/src/main/java/org/elasticsearch/action/fieldcaps/RequestDispatcher.java:
```diff
@@ -127,31 +134,27 @@ protected void doRun() {
-                for (ShardRouting shard : selectedShards) {
-                    nodeToSelectedShards.computeIfAbsent(shard.currentNodeId(), n -> new ArrayList<>()).add(shard.shardId());
+        groups.values().removeIf(g -> g.completed.get());
+        final Set<Group> visited = Collections.newSetFromMap(new IdentityHashMap<>());
```
Yes, HashSet should be good. I used IdentityHashMap as I worried that someone could accidentally implement hashCode and equals for Group.

dnhatn committed:
Merge remote-tracking branch 'elastic/master' into field_caps_groups

dnhatn committed:
Fix ResponseRewriter

dnhatn committed:
Simplify

dnhatn committed:
shared response

dnhatn committed:
Add TODO

dnhatn commented:
@original-brownbear Thank you for running the benchmark. Did you run against all indices or only `auditbeat*` indices? I see a small improvement (7%) with `auditbeat*` indices only.

```
| 50th percentile latency |              field-caps |   4576.27        |   4254.51        |   -321.753   |     ms |    -7.03% |
| 90th percentile latency |              field-caps |   5482.8         |   4805.08        |   -677.729   |     ms |   -12.36% |
| 99th percentile latency |              field-caps |   6001.41        |   5248.21        |   -753.198   |     ms |   -12.55% |
|100th percentile latency |              field-caps |   6241.83        |   5353.3         |   -888.527   |     ms |   -14.24% |
```

dnhatn committed:
shared response

dnhatn commented:
@jtibshirani @original-brownbear Thank you for your reviews. I think I have addressed your comments.

> I wonder if Kibana is using index_filter or not, this will really affect when the optimization helps! Unfortunately I think it would require a big change to how we execute field caps in order to get it to work with index_filter...

One of the options is to add the `can_match` phase for requests with index_filter to eliminate unmatched copies as we discussed. We should have a single execution mode in RequestDispatcher after that. I can do this in a follow-up.

dnhatn commented:
The optimization introduced in this PR doesn't reduce memory usage or latency of field-caps requests. I will close this PR and try to get https://github.com/elastic/elasticsearch/pull/86323 in instead. Thanks everyone for reviewing.
